node/primitives/src/lib.rs

```diff
@@ -299,8 +302,8 @@ pub struct Proof(BoundedVec<BoundedVec<u8, 1, MERKLE_NODE_MAX_SIZE>, 1, MERKLE_P
 
 impl Proof {
 	/// This function allows to convert back to the standard nested Vec format
-	pub fn as_vec(&self) -> Vec<Vec<u8>> {
-		self.0.as_vec().iter().map(|v| v.as_vec().clone()).collect()
+	pub fn as_vec(&self) -> Vec<&[u8]> {
```

Comment:
I would like to modify this to return an `&[&[_]]`, but everything I tried required to call `collect()`, at which point the overhead is equivalent to using a `Vec<_>`, no?

Reply:
`fn iter(&self) -> impl Iterator<Item = &[u8]>`

Reply:
```patch
diff --git a/erasure-coding/src/lib.rs b/erasure-coding/src/lib.rs
index 6a566bf1f..4ba8350ac 100644
--- a/erasure-coding/src/lib.rs
+++ b/erasure-coding/src/lib.rs
@@ -300,7 +300,7 @@ where
 /// index.
 pub fn branch_hash(root: &H256, branch_nodes: &Proof, index: usize) -> Result<H256, Error> {
 	let mut trie_storage: MemoryDB<Blake2Hasher> = MemoryDB::default();
-	for node in branch_nodes.as_vec().iter() {
+	for node in branch_nodes.iter() {
 		(&mut trie_storage as &mut trie::HashDB<_>).insert(EMPTY_PREFIX, node);
 	}
 
diff --git a/node/primitives/src/lib.rs b/node/primitives/src/lib.rs
index edfd14b66..6f9c6fc46 100644
--- a/node/primitives/src/lib.rs
+++ b/node/primitives/src/lib.rs
@@ -302,8 +302,8 @@ pub struct Proof(BoundedVec<BoundedVec<u8, 1, MERKLE_NODE_MAX_SIZE>, 1, MERKLE_P
 
 impl Proof {
 	/// This function allows to convert back to the standard nested Vec format
-	pub fn as_vec(&self) -> Vec<&[u8]> {
-		self.0.iter().map(|v| v.as_slice()).collect::<Vec<&[u8]>>()
+	pub fn iter(&self) -> impl Iterator<Item = &[u8]> {
+		self.0.iter().map(|v| v.as_slice())
 	}
 }
 
@@ -359,7 +359,7 @@ impl Encode for Proof {
 	}
 
 	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
-		let temp = self.as_vec();
+        let temp = self.0.iter().map(|v| v.as_vec()).collect::<Vec<_>>();
 		temp.using_encoded(f)
 	}
 }
```